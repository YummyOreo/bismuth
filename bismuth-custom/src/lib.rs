#![allow(dead_code, unused)]
use std::collections::HashMap;

use bismuth_parser::{
    custom::CustomElm,
    tree::{Ast, Element, Kind},
    Metadata, Parser,
};

mod builtin;
pub mod plugin;
pub mod template;

pub struct Custom {
    name: String,
    data: HashMap<String, String>,
    template: Option<template::Template>,
    plugin: Option<plugin::Plugin>,
}

impl Custom {
    pub fn new(name: String, data: HashMap<String, String>) -> Self {
        Custom {
            name,
            data,
            template: None,
            plugin: None,
        }
    }

    pub fn from_elm(elm: &CustomElm) -> Self {
        Self::new(elm.name.clone(), elm.values.clone())
    }

    fn find_plugin(&mut self) -> Option<plugin::Plugin> {
        // REDO THIS WHEN YOU IMPLEMENT PLUGINS
        match self.name.to_string().as_str() {
            "blog list" | "blogs" | "bloglist" => {}
            "navbar" => {}
            "footer" => {}
            _ => {}
        }
        None
    }
    fn find_template(&mut self) -> Option<template::Template> {
        // REDO THIS WHEN YOU IMPLEMENT TEMPLATE
        match self.name.to_string().as_str() {
            "blog list" | "blogs" | "bloglist" => {}
            "navbar" => {}
            "footer" => {}
            _ => {}
        }
        None
    }

    pub fn find(&mut self) {
        self.plugin = self.find_plugin();
        self.template = self.find_template();
    }
}

fn get_customs(ast: &Ast) -> Vec<&Element> {
    let elements: Vec<&Element> = ast.elements.iter().collect();
    elements
        .iter()
        .filter_map(|&e| {
            if let Kind::CustomElement(_) = &e.kind {
                Some(e)
            } else {
                None
            }
        })
        .collect::<Vec<&Element>>()
}

pub fn parse_custom(mut target: Parser, others: Vec<&Parser>) -> Parser {
    let custom_elms = get_customs(&target.ast);
    let customs: Vec<Custom> = custom_elms
        .iter()
        .filter_map(|e| {
            if let Kind::CustomElement(c) = &e.kind {
                let mut custom = Custom::from_elm(c);
                custom.find();
                Some(custom)
            } else {
                None
            }
        })
        .collect();
    todo!()
}

#[cfg(test)]
mod test_utils {
    use super::*;
    use regex::Regex;

    macro_rules! snapshot {
        ($content:tt) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!($content);
            });
        };
    }
    #[test]
    fn get_customs_test() {
        let mut parser =
            bismuth_parser::Parser::new_test("/test/", "%{{\nname: test\nother: key\n}}");
        parser.parse().unwrap();

        let customs = format!("{:#?}", get_customs(&parser.ast));
        let re = Regex::new(r"id: \d+").unwrap();
        let customs = re.replace_all(&customs, "id: [redacted]").to_string();
        snapshot!(customs);
    }
}
