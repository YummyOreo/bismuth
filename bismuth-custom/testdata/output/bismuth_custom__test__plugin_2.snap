---
source: bismuth-custom/src/lib.rs
expression: "snapshot(\"%{{\\nname: bloglist\\nother: key\\n}}\")"
---
FrontMatter {
    title: Some(
        "test",
    ),
    path: Some(
        "/test",
    ),
    kind: Some(
        "default",
    ),
    values: None,
}
Element{
    Kind: "Custom{Name: bloglist, Body: None, Values: [(\"other\", \"key\")], Template: None}",
    Text: None,
    Attrs: [],
    Elements: [
    Element{
        Kind: "Custom{Name: builtin_blogwrapper, Body: None, Values: [], Template: Some(\"<style>\\r\\n    .blogitem {\\r\\n        width: 9rem;\\r\\n        background-color: var(--background-2);\\r\\n        padding: 0.5rem 0.5rem;\\r\\n        border-radius: 0.5rem;\\r\\n    }\\r\\n\\r\\n    .blogitem-title {\\r\\n        margin-bottom: 0rem;\\r\\n        margin-top: 0.5rem;\\r\\n    }\\r\\n\\r\\n    .blogitem-date {\\r\\n        font-size: 0.7rem;\\r\\n        color: var(--text-22);\\r\\n    }\\r\\n\\r\\n    .blogitem-button {\\r\\n        border: none;\\r\\n        border-radius: 0.5rem;\\r\\n        height: 2rem;\\r\\n        width: 4rem;\\r\\n        margin-left: 50%;\\r\\n\\r\\n        background-color: var(--link);\\r\\n    }\\r\\n\\r\\n    .blogitem-button::hover {\\r\\n        background-color: var(--link-hover);\\r\\n    }\\r\\n\\r\\n    .blog-wrapper {\\r\\n        width: 10rem;\\r\\n        padding: 1rem 1rem;\\r\\n        background-color: var(--background-3);\\r\\n        border-radius: 0.5rem;\\r\\n    }\\r\\n</style>\\r\\n<div class=\\\"blog-wrapper\\\">\\r\\n    {elements}\\r\\n</div>\\r\\n\")}",
        Text: None,
        Attrs: [],
        Elements: [
        ])
    },
    ])
}
Element{
    Kind: "EndOfLine",
    Text: None,
    Attrs: [],
    Elements: [
    ])
}

